### ISSUE TITLE
Snippet to dump binary data and not to xml

### ISSUE THREAD

**Issue by agarciagonzalez** [fonttools/fonttools#1143]
Hi
How can I dump or export to manipulate later data of an specific table or tag?

f.saveXML(0, newlinestr='\n', tables=['Glat'])

Generates a XML containing hex data, but I wish I could manipulate that data without Xml. Also once binary data is manipulated I would like to import it into my font with a valid checksum without Xml again.

Example:

Open font
for tag in ..
checkSum
do something with data of table Glat
fix checksum
check font is OK
save font to file

Thanks for the help

**Comment by jenskutilek** [comment]
You can get the raw table data like this:

```python
font = TTFont("myfont.ttf")
table_tag = "Glat"
data = font.getTableData(table_tag)
```

To assign the manipulated data to a font table:

```python
from fontTools.ttLib.tables.DefaultTable import DefaultTable

target_font = TTFont("myfont.ttf")
table_tag = "Glat"
table = DefaultTable(table_tag)
table.data = manipulated_data
target_font[table_tag] = table
target_font.save("myfont_modified.ttf")
```

You don't need to calculate the checksum, fontTools will do it automatically.

**Comment by anthrotype** [comment]
thanks @jenskutilek. That could be a nice script to have in Snippets.

@agarciagonzalez I wonder, don't we already support dumping/merging Glat table (a Graphite2 table) to/from TTX?

https://github.com/fonttools/fonttools/blob/437b3ead2e0921af691679b8207b4c18912e7b7c/Lib/fontTools/ttLib/tables/G__l_a_t.py

If the user is getting hex data it might be an issue while decompiling, and fonttools is falling back to the DefaultTable implementation which prints raw hexadecimal data.

@agarciagonzalez are you getting any error or warning when you dump your font to TTX?

cc @mhosken

**Comment by agarciagonzalez** [comment]
Works like a charm, thank you

**Comment by jenskutilek** [comment]
I've added a snippet: https://github.com/fonttools/fonttools/commit/0cae12bdbde6fa87f65025bc14ee5863f2526c4b

**Comment by agarciagonzalez** [comment]
is there any way that if any table contains an incorrect checksum fix all? Graphite2 tables. Assuming that I do not know which of the tables the checksum may be incorrect.

**Comment by mhosken** [comment]
Why Graphite2 tables in particular?

**Comment by agarciagonzalez** [comment]
Because is mentioned before as example

**Comment by anthrotype** [comment]
I didn’t understand the question, could you rephrase it?

**Comment by agarciagonzalez** [comment]
Something like:
1 Open file
check tables checksum (Glat or other corrupted table checksum, before I only modified tables using fonttools but now also using another program so I've to update the checksum of the table if is wrong)

2 fix checksum
3 check font is OK
4 save font to file

**Comment by anthrotype** [comment]
That’s done automatically when you save a font with fonttools. Each sfnt table checksum and the checksum adjustment in the head table are always recomputed by the SFNTWriter, when yo do TTFont.save

**Comment by agarciagonzalez** [comment]
Just I need to open and close the font and fixes the corrupt tables?

**Comment by anthrotype** [comment]
Just the sfnt header is “fixed”. If the table data was corrupt beyond the checksum then it will stay corrupt

**Comment by agarciagonzalez** [comment]
Then how can I fix the Glat checksum for example?

**Comment by anthrotype** [comment]
I don’t know what you mean by “Glat checksum”. If you mean the checksum that is stored in the sfnt font file header (the global list of sfnt table entries, with tag, length, offset and checksum) then yes, it will be recalculated to the correct one every time you save a TTFont.

**Comment by anthrotype** [comment]
By the way, the snippet that @jenskutilek shared above already does that, you don’t need to do anything else to fix the sfnt table checksums.

**Comment by agarciagonzalez** [comment]
It fixes the table checksum, but knowing in advance which table is modified because it's modified within fonttools, but if is a full font modified with an external program, my question was: just opening and saving the font file will update the whole checksum to the tables to work?

**Comment by anthrotype** [comment]
the checksum is not stored inside the table itself, it's stored in the sftn header which contains a list entries for each of the table, that's where checksum (along with tag, length, offset) are. All the sfnt header (plus the checksumAdjustment field and optionally the modified timestamp in the head table) are recomputed every time you save a font. So yes.

**Comment by agarciagonzalez** [comment]
Open and font.save is not working
```
'created' timestamp out of range; ignoring top bytes
'created' timestamp seems very low; regarding as unix timestamp
'modified' timestamp out of range; ignoring top bytes
'modified' timestamp seems very low; regarding as unix timestamp
Traceback (most recent call last):
  File "opensave.py", line 35, in <module>
    font.save(fontname)
  File "/usr/lib/python2.7/site-packages/fontTools/ttLib/ttFont.py", line 165, in save
    writer_reordersTables = self._save(tmp)
  File "/usr/lib/python2.7/site-packages/fontTools/ttLib/ttFont.py", line 204, in _save
    self._writeTable(tag, writer, done, tableCache)
  File "/usr/lib/python2.7/site-packages/fontTools/ttLib/ttFont.py", line 619, in _writeTable
    tabledata = self.getTableData(tag)
  File "/usr/lib/python2.7/site-packages/fontTools/ttLib/ttFont.py", line 640, in getTableData
    return self.reader[tag]
  File "/usr/lib/python2.7/site-packages/fontTools/ttLib/sfnt.py", line 103, in __getitem__
    data = entry.loadData (self.file)
  File "/usr/lib/python2.7/site-packages/fontTools/ttLib/sfnt.py", line 453, in loadData
    assert len(data) == self.length
AssertionError
```

**Comment by anthrotype** [comment]
opening the font is not working, because you probably modified a table's data with some external tool without also updating the length field for that table in the sfnt header (the latter is also called a "Offset Table" in the opentype spec).

Why don't you use the method suggested by Jens to inject an arbitrary table data in an existing TTFont object letting fonttools compile a correct sfnt header?
